### PR TITLE
4886-speedup solr indexing

### PR DESCRIPTION
**Pull request by pkdash**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. The speed grain during full re-index can be tested only by deploying this work on beta. 
2. However, it would be good to verify that the discover works as before. Create a Resource and make it discoverable. Verify, that the resource is listed in discover. Verify the items listed under each Filter category looks correct. Add a new keyword and a new author (make this author the first author) to the resource. Verify the new author and the new keyword are listed in their respective filters. You can repeat this for a Collection and App Connector.

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4901](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4901)

File: hs_core/search_indexes.py [hunk 1]
```diff
@@ -711,67 +692,82 @@ def prepare_geometry_type(self, obj):
...
+            for f in obj.cached_reftimeseries_logical_files:
+                for v in f.metadata.variables:
+                    # TODO: inconsistent use of variable code and variable name
+                    if discoverable(v.name):
+                        variables.add(v.name.strip())
+            for f in obj.georasterlogicalfile_set.all():
+                for b in f.metadata.bandInformations:
+                    if discoverable(b.variableName):
+                        variables.add(b.variableName)
         return list(variables)
```
Comment (devincowan) [inline hs_core/search_indexes.py:771]: return None if not a Composite resource?
same question for all of the remaining functions in this Class -- should they default return None?

**Review by devincowan — APPROVED** [review]
Looks good! A few questions for you...

File: hs_core/search_indexes.py [hunk 2]
```diff
@@ -303,80 +335,85 @@ def prepare_author_url(self, obj):
...
-        if hasattr(obj, 'metadata') and \
-                obj.metadata is not None and \
-                obj.metadata.subjects is not None:
-            return [subject.value.strip() for subject in obj.metadata.subjects.all()
-                    .exclude(value__isnull=True)]
-        else:
-            return []
+        subjects = []
+        if self._has_metadata(obj):
+            for subject in obj.metadata.subjects.all():
```
[the remark this answers is not in the document]
Comment (pkdash) [reply hs_core/search_indexes.py:391]: metadata subjects (keywords) only accessed once during the resource index. We are caching querysets that are accessed more than once.

File: hs_core/search_indexes.py [hunk 3]
```diff
@@ -229,16 +232,41 @@ class BaseResourceIndex(indexes.SearchIndex, indexes.Indexable):
...
+        return hasattr(obj, 'metadata') and obj.metadata is not None
+
+    def _get_coverage_bounding_box_value(self, obj, box_direction):
+        """Gets the coverage value for the box limit
+        :param  obj: an instance of resource
+        :param  box_direction: One of these: northlimit, westlimit, southlimit, or eastlimit
+        """
+        if self._has_metadata(obj):
+            self._cache_queryset(obj, 'cached_coverages', obj.metadata.coverages.all())
+            # TODO: does not index properly if there are multiple coverages of the same type.
```
[the remark this answers is not in the document]
Comment (pkdash) [reply hs_core/search_indexes.py:251]: The 'TODO' comment was probably by Alva and I don't understand it. Why would be there multiple coverage of the same type?

File: hs_core/search_indexes.py (hunk 1 above)
Comment (pkdash) [reply hs_core/search_indexes.py:771]: It seems we need to return an empty list when more than one item is expected.

File: hs_core/search_indexes.py [hunk 4]
```diff
@@ -248,11 +276,12 @@ def prepare_modified(self, obj):
...
     def prepare_title(self, obj):
         """Return metadata title if exists, otherwise return 'none'."""
-        if hasattr(obj, 'metadata') and \
-                obj.metadata is not None and \
-                obj.metadata.title is not None and \
-                obj.metadata.title.value is not None:
-            return obj.metadata.title.value.lstrip()
+
+        if self._has_metadata(obj):
+            self._cache_queryset(obj, 'cached_title', obj.metadata.title)
```
[the remark this answers is not in the document]
Comment (pkdash) [reply hs_core/search_indexes.py:281]: You are right. I need to fix that.

[the remark this answers is not in the document]
Comment (pkdash) [reply hs_core/search_indexes.py]: fixed.

File: hs_core/search_indexes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (devincowan) [reply hs_core/search_indexes.py:392]: makes sense

File: hs_core/search_indexes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (devincowan) [reply hs_core/search_indexes.py:251]: Sounds good, I'm fine to ignore this or remove it

File: hs_core/search_indexes.py (hunk 1 above)
Comment (devincowan) [reply hs_core/search_indexes.py:772]: Ah nice, okay!

**Review by devincowan — APPROVED** [review]

**Pull request merged**